src/poh_recorder.rs

```diff
@@ -0,0 +1,104 @@
+//! The `poh_service` module provides an object for generating a Proof of History.
+//! It records Hashes items on behalf of its users. It continuously generates
+//! new Hashes, only stopping to check if it has been sent a Hash to mix in
+//! to the Poh.
+//!
+//! The returned Entry includes the mix-in request, the latest Poh Hash, and the
+//! number of Hashes generated in the service since the last mix-in request.
+//!
+//! The resulting stream of Hashes represents ordered events in time.
+//!
+use bank::Bank;
+use entry::Entry;
+use hash::Hash;
+use poh::Poh;
+use result::Result;
+use std::sync::mpsc::Sender;
+use std::sync::{Arc, Mutex};
+use transaction::Transaction;
+
+#[derive(Clone)]
+pub struct PohRecorder {
+    poh: Arc<Mutex<Poh>>,
+    bank: Arc<Bank>,
+    sender: Sender<Vec<Entry>>,
+}
+
+impl PohRecorder {
+    /// A recorder to synchronize PoH with the following data structures
+    /// * bank - the LastId's queue is updated on `tick` and `record` events
+    /// * sender - the Entry channel that outputs to the ledger
+    pub fn new(bank: Arc<Bank>, sender: Sender<Vec<Entry>>) -> Self {
+        let poh = Arc::new(Mutex::new(Poh::new(bank.last_id())));
+        PohRecorder { poh, bank, sender }
+    }
+
+    pub fn hash(&self) {
+        // TODO: amortize the cost of this lock by doing the loop in here for
+        // some min amount of hashes
+        let mut poh = self.poh.lock().unwrap();
+        poh.hash()
+    }
+
+    pub fn tick(&self) -> Result<()> {
+        // Register and send the entry out while holding the lock.
+        // This guarantees PoH order and Entry production and banks LastId queue is the same
+        let mut poh = self.poh.lock().unwrap();
+        let tick = poh.tick();
+        self.bank.register_entry_id(&tick.id);
+        let entry = Entry {
+            num_hashes: tick.num_hashes,
+            id: tick.id,
+            transactions: vec![],
+        };
+        self.sender.send(vec![entry])?;
+        Ok(())
+    }
+
+    pub fn record(&self, mixin: Hash, txs: Vec<Transaction>) -> Result<()> {
+        // Register and send the entry out while holding the lock.
+        // This guarantees PoH order and Entry production and banks LastId queue is the same.
+        let mut poh = self.poh.lock().unwrap();
```

Comment:
you wanna drop this lock fast

Reply:
@rob-solana we would need two locks then, one for poh itself, the other one to sync last_id's and entry output queue.  we can fix that later.  i don't think we need a lock around poh itself

Reply:
this lock guards sender.send()?